File: php/tree/Tree.php
<?php declare(strict_types=1);

namespace bobel\tree;

class Tree
{
    public ?Node $root = null;

    public function traverseBF(callable $fn): void
    {
        $list = [$this->root];

        while (count($list)) {
            $node = array_shift($list);
            array_push($list, ...$node->children);

            $fn($node);
        }
    }

    public function traverseDF(callable $fn): void
    {
        $list = [$this->root];

        while (count($list)) {
            $node = array_shift($list);
            $list = array_merge($node->children, $list);

            $fn($node);
        }
    }
}

File: csharp/tree/Tree.cs
using System.Collections.Generic;

namespace Tree
{
    public delegate void ProcessNode(Node node);

    public class Tree
    {
        public Node Root = null;

        public void TraverseBF(ProcessNode function)
        {
            var list = new List<Node>();
            list.Add(this.Root);

            while (list.Count > 0)
            {
                Node node = list[0];
                list.RemoveAt(0);

                list.AddRange(node.Children);

                function(node);
            }
        }

        public void TraverseDF(ProcessNode function)
        {
            var list = new List<Node>();
            list.Add(this.Root);

            while (list.Count > 0)
            {
                Node node = list[0];
                list.RemoveAt(0);

                list.InsertRange(0, node.Children);

                function(node);
            }
        }
    }
}